# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [3]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [4]:
view_sentence_range = (12, 20)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 12 to 20:
india is rainy during june , and it is sometimes warm in november .
that cat was my most loved animal .
he dislikes grapefruit , limes , and lemons .
her least liked fruit is the lemon , but his least liked is the grapefruit .
california is never cold during february , but it is sometimes freezing in june .
china is usually pleasant during autumn , and it is usually quiet in october .
paris is never freezing during november , but it is wonderful in october .
the united states is never rainy during january , but it is sometimes mild in october .

French sentences 12 to 20:
inde est pluvieux en juin , et il est parfois chaud en novembre .
ce chat était mon animal le plus aimé .
il n'aime pamplemousse , citrons verts et les citrons .
son fruit est moins aimé le citron , mais son moins aimé est le pamplemousse .
californie 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [5]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param showource_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # spit line-by-line since each sentence is on separate line
    source_words = source_text.split('\n')
    target_words = target_text.split('\n')
#     print(len(source_text))
#     print(source_words[:5])
#     print(target_words[:5])
    
    # for each line convert words into array of Int
    source_ids = [[source_vocab_to_int[word] for word in words.split()] for words in source_words]
    target_ids = [[target_vocab_to_int[word] for word in words.split()] for words in target_words]
    
    # add EOS int to end of targets only
    target_ids = [target + [target_vocab_to_int['<EOS>']] for target in target_ids]
    
#     print(len(source_ids))
#     print(len(target_ids))
    return source_ids, target_ids

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [7]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper
import problem_unittests as tests

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [8]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.1.0
Default GPU Device: /gpu:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [10]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    _input = tf.placeholder(tf.int32, [None, None], name = "input")
    _targets = tf.placeholder(tf.int32, [None, None], name = "targets")
    
    _learning_rate = tf.placeholder(tf.float32, name ="learning_rate")
    _keep_prob = tf.placeholder(tf.float32, name = "keep_prob")
    
    _target_seq_len = tf.placeholder(tf.int32,[None], name = "target_sequence_length")
    _max_target_seq_len =  tf.reduce_max(_target_seq_len, name = "max_target_len")
    _source_target_seq_len = tf.placeholder(tf.int32, [None], name = "source_sequence_length")
    
    return _input, _targets, _learning_rate, _keep_prob, _target_seq_len, _max_target_seq_len, _source_target_seq_len


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [11]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    popped_last_slice = tf.strided_slice(target_data, begin=[0,0], end=[batch_size, -1], strides=[1,1])
    go_inputs = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']), popped_last_slice], 1)
    
    return go_inputs

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [13]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    embeded = tf.contrib.layers.embed_sequence(rnn_inputs, source_vocab_size, encoding_embedding_size)
    cell_initializer = tf.random_uniform_initializer(-0.1, 0.1, seed=2)
    make_cell = lambda size: tf.contrib.rnn.LSTMCell(size,initializer=cell_initializer)
    cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size) for _ in range(num_layers)])
    cell = tf.contrib.rnn.DropoutWrapper(cell, input_keep_prob=keep_prob)
    output, state = tf.nn.dynamic_rnn(cell,embeded,sequence_length=source_sequence_length, dtype=tf.float32)

    # TODO: Implement Function
    return output, state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [14]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    
    helper = tf.contrib.seq2seq.TrainingHelper(inputs=dec_embed_input,
                                               sequence_length=target_sequence_length, 
                                               time_major=False)

    
    decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell,
                                              helper, 
                                              encoder_state,
                                              output_layer)

    output, _ = tf.contrib.seq2seq.dynamic_decode(
        decoder=decoder,
        output_time_major=False,
        impute_finished=True,
        maximum_iterations=max_summary_length)
        
    return output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [15]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    go_constant = tf.constant([start_of_sequence_id], dtype=tf.int32)
    start_token = tf.tile(go_constant,[batch_size], name="start_token")
    helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings,
                                                      start_token,
                                                      end_of_sequence_id)
    
    decoder = tf.contrib.seq2seq.BasicDecoder(cell=dec_cell,
                                              helper=helper,
                                              initial_state=encoder_state,
                                              output_layer=output_layer)
    
    output, _ = tf.contrib.seq2seq.dynamic_decode(decoder=decoder, 
                                                  impute_finished=True, 
                                                  maximum_iterations=max_target_sequence_length)
    return output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [16]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """

    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)
    
    cell_initializer = tf.random_uniform_initializer(-0.1, 0.1, seed=2)
    make_cell = lambda size: tf.contrib.rnn.LSTMCell(size,initializer=cell_initializer)
    cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size) for _ in range(num_layers)])
    cell = tf.contrib.rnn.DropoutWrapper(cell, input_keep_prob=keep_prob)
    
    output_layer = Dense(target_vocab_size,
                         kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1))

    kernal_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1)
    output_layer = Dense(target_vocab_size,kernel_initializer=kernal_initializer)
    
    with tf.variable_scope("decode"):
        train_logits = decoding_layer_train(encoder_state=encoder_state,
                                            dec_cell=cell,
                                            dec_embed_input=dec_embed_input,
                                            target_sequence_length=target_sequence_length,
                                            max_summary_length=max_target_sequence_length,
                                            output_layer=output_layer,
                                            keep_prob=keep_prob)

    with tf.variable_scope("decode", reuse=True):
        infer_logits = decoding_layer_infer(encoder_state=encoder_state,
                                            dec_cell=cell,
                                            dec_embeddings=dec_embeddings,
                                            start_of_sequence_id=target_vocab_to_int['<GO>'], 
                                            end_of_sequence_id=target_vocab_to_int['<EOS>'],
                                            max_target_sequence_length=max_target_sequence_length, 
                                            vocab_size=target_vocab_size,
                                            output_layer=output_layer, 
                                            batch_size=batch_size,
                                            keep_prob=keep_prob)
    return train_logits, infer_logits



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [17]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    enc_output, enc_state = encoding_layer(input_data,
                                   rnn_size,
                                   num_layers,
                                   keep_prob,
                                   source_sequence_length,
                                   source_vocab_size,
                                   enc_embedding_size)
    
    dec_input = process_decoder_input(target_data,
                                       target_vocab_to_int,
                                       batch_size)
    
    training, inference = decoding_layer(dec_input,
                                         enc_state,
                                         target_sequence_length,
                                         max_target_sentence_length,
                                         rnn_size,
                                         num_layers,
                                         target_vocab_to_int,
                                         target_vocab_size,
                                         batch_size,
                                         keep_prob,
                                         dec_embedding_size)
    return training, inference


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [76]:
# Number of Epochs
epochs = 9
# Batch Size
batch_size = 128
# RNN Size
rnn_size = 512
# Number of Layers
num_layers = 3
# Embedding Size
encoding_embedding_size = 56
decoding_embedding_size = 56
# Learning Rate
learning_rate = 0.005
# Dropout Keep Probability
keep_probability = 0.67
display_step = 5

### Build the Graph
Build the graph using the neural network you implemented.

In [19]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [20]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [77]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch    5/1077 - Train Accuracy: 0.2605, Validation Accuracy: 0.3050, Loss: 4.5787
Epoch   0 Batch   10/1077 - Train Accuracy: 0.1982, Validation Accuracy: 0.3050, Loss: 3.9466
Epoch   0 Batch   15/1077 - Train Accuracy: 0.2707, Validation Accuracy: 0.3363, Loss: 3.5186
Epoch   0 Batch   20/1077 - Train Accuracy: 0.2699, Validation Accuracy: 0.3370, Loss: 3.3823
Epoch   0 Batch   25/1077 - Train Accuracy: 0.2855, Validation Accuracy: 0.3434, Loss: 3.3716
Epoch   0 Batch   30/1077 - Train Accuracy: 0.2672, Validation Accuracy: 0.3366, Loss: 3.2338
Epoch   0 Batch   35/1077 - Train Accuracy: 0.3078, Validation Accuracy: 0.3683, Loss: 3.1666
Epoch   0 Batch   40/1077 - Train Accuracy: 0.3121, Validation Accuracy: 0.3750, Loss: 3.1080
Epoch   0 Batch   45/1077 - Train Accuracy: 0.3152, Validation Accuracy: 0.3789, Loss: 3.0424
Epoch   0 Batch   50/1077 - Train Accuracy: 0.3078, Validation Accuracy: 0.3778, Loss: 3.0306
Epoch   0 Batch   55/1077 - Train Accuracy: 0.3578, Validati

Epoch   0 Batch  445/1077 - Train Accuracy: 0.4934, Validation Accuracy: 0.5398, Loss: 1.0429
Epoch   0 Batch  450/1077 - Train Accuracy: 0.4875, Validation Accuracy: 0.5270, Loss: 0.9847
Epoch   0 Batch  455/1077 - Train Accuracy: 0.5344, Validation Accuracy: 0.5391, Loss: 0.9071
Epoch   0 Batch  460/1077 - Train Accuracy: 0.4688, Validation Accuracy: 0.5369, Loss: 0.9911
Epoch   0 Batch  465/1077 - Train Accuracy: 0.4671, Validation Accuracy: 0.5380, Loss: 1.0337
Epoch   0 Batch  470/1077 - Train Accuracy: 0.4679, Validation Accuracy: 0.5511, Loss: 0.9915
Epoch   0 Batch  475/1077 - Train Accuracy: 0.5336, Validation Accuracy: 0.5462, Loss: 0.9003
Epoch   0 Batch  480/1077 - Train Accuracy: 0.5156, Validation Accuracy: 0.5419, Loss: 0.9546
Epoch   0 Batch  485/1077 - Train Accuracy: 0.5488, Validation Accuracy: 0.5401, Loss: 0.9076
Epoch   0 Batch  490/1077 - Train Accuracy: 0.4941, Validation Accuracy: 0.5312, Loss: 0.9393
Epoch   0 Batch  495/1077 - Train Accuracy: 0.5023, Validati

Epoch   0 Batch  885/1077 - Train Accuracy: 0.6218, Validation Accuracy: 0.5845, Loss: 0.6013
Epoch   0 Batch  890/1077 - Train Accuracy: 0.6343, Validation Accuracy: 0.5827, Loss: 0.6433
Epoch   0 Batch  895/1077 - Train Accuracy: 0.5687, Validation Accuracy: 0.5845, Loss: 0.6754
Epoch   0 Batch  900/1077 - Train Accuracy: 0.5633, Validation Accuracy: 0.5646, Loss: 0.6887
Epoch   0 Batch  905/1077 - Train Accuracy: 0.6105, Validation Accuracy: 0.5934, Loss: 0.6558
Epoch   0 Batch  910/1077 - Train Accuracy: 0.5681, Validation Accuracy: 0.5927, Loss: 0.6676
Epoch   0 Batch  915/1077 - Train Accuracy: 0.5415, Validation Accuracy: 0.5877, Loss: 0.7217
Epoch   0 Batch  920/1077 - Train Accuracy: 0.5504, Validation Accuracy: 0.5771, Loss: 0.6813
Epoch   0 Batch  925/1077 - Train Accuracy: 0.5644, Validation Accuracy: 0.5984, Loss: 0.6630
Epoch   0 Batch  930/1077 - Train Accuracy: 0.5836, Validation Accuracy: 0.5991, Loss: 0.6606
Epoch   0 Batch  935/1077 - Train Accuracy: 0.5684, Validati

Epoch   1 Batch  250/1077 - Train Accuracy: 0.5845, Validation Accuracy: 0.6154, Loss: 0.5244
Epoch   1 Batch  255/1077 - Train Accuracy: 0.5934, Validation Accuracy: 0.6069, Loss: 0.5446
Epoch   1 Batch  260/1077 - Train Accuracy: 0.6071, Validation Accuracy: 0.6214, Loss: 0.5225
Epoch   1 Batch  265/1077 - Train Accuracy: 0.5738, Validation Accuracy: 0.6218, Loss: 0.5698
Epoch   1 Batch  270/1077 - Train Accuracy: 0.5961, Validation Accuracy: 0.6211, Loss: 0.5716
Epoch   1 Batch  275/1077 - Train Accuracy: 0.6064, Validation Accuracy: 0.6364, Loss: 0.5173
Epoch   1 Batch  280/1077 - Train Accuracy: 0.6109, Validation Accuracy: 0.6246, Loss: 0.5644
Epoch   1 Batch  285/1077 - Train Accuracy: 0.6228, Validation Accuracy: 0.6268, Loss: 0.5198
Epoch   1 Batch  290/1077 - Train Accuracy: 0.5969, Validation Accuracy: 0.6587, Loss: 0.5560
Epoch   1 Batch  295/1077 - Train Accuracy: 0.6238, Validation Accuracy: 0.6268, Loss: 0.5675
Epoch   1 Batch  300/1077 - Train Accuracy: 0.5855, Validati

Epoch   1 Batch  690/1077 - Train Accuracy: 0.6832, Validation Accuracy: 0.6417, Loss: 0.4497
Epoch   1 Batch  695/1077 - Train Accuracy: 0.6707, Validation Accuracy: 0.6523, Loss: 0.4248
Epoch   1 Batch  700/1077 - Train Accuracy: 0.6520, Validation Accuracy: 0.6413, Loss: 0.4370
Epoch   1 Batch  705/1077 - Train Accuracy: 0.6443, Validation Accuracy: 0.6488, Loss: 0.4837
Epoch   1 Batch  710/1077 - Train Accuracy: 0.6477, Validation Accuracy: 0.6474, Loss: 0.4607
Epoch   1 Batch  715/1077 - Train Accuracy: 0.6773, Validation Accuracy: 0.6634, Loss: 0.4600
Epoch   1 Batch  720/1077 - Train Accuracy: 0.6678, Validation Accuracy: 0.6559, Loss: 0.4722
Epoch   1 Batch  725/1077 - Train Accuracy: 0.6767, Validation Accuracy: 0.6626, Loss: 0.4177
Epoch   1 Batch  730/1077 - Train Accuracy: 0.6527, Validation Accuracy: 0.6705, Loss: 0.4418
Epoch   1 Batch  735/1077 - Train Accuracy: 0.6859, Validation Accuracy: 0.6410, Loss: 0.4351
Epoch   1 Batch  740/1077 - Train Accuracy: 0.6687, Validati

Epoch   2 Batch   55/1077 - Train Accuracy: 0.7039, Validation Accuracy: 0.6818, Loss: 0.3564
Epoch   2 Batch   60/1077 - Train Accuracy: 0.7169, Validation Accuracy: 0.6783, Loss: 0.3487
Epoch   2 Batch   65/1077 - Train Accuracy: 0.6780, Validation Accuracy: 0.6896, Loss: 0.3788
Epoch   2 Batch   70/1077 - Train Accuracy: 0.6764, Validation Accuracy: 0.6680, Loss: 0.3976
Epoch   2 Batch   75/1077 - Train Accuracy: 0.6820, Validation Accuracy: 0.6722, Loss: 0.3883
Epoch   2 Batch   80/1077 - Train Accuracy: 0.6602, Validation Accuracy: 0.6712, Loss: 0.3723
Epoch   2 Batch   85/1077 - Train Accuracy: 0.7305, Validation Accuracy: 0.6829, Loss: 0.3551
Epoch   2 Batch   90/1077 - Train Accuracy: 0.6672, Validation Accuracy: 0.6882, Loss: 0.3788
Epoch   2 Batch   95/1077 - Train Accuracy: 0.7117, Validation Accuracy: 0.6960, Loss: 0.3658
Epoch   2 Batch  100/1077 - Train Accuracy: 0.6766, Validation Accuracy: 0.7063, Loss: 0.3669
Epoch   2 Batch  105/1077 - Train Accuracy: 0.6996, Validati

Epoch   2 Batch  495/1077 - Train Accuracy: 0.6871, Validation Accuracy: 0.6921, Loss: 0.3109
Epoch   2 Batch  500/1077 - Train Accuracy: 0.7250, Validation Accuracy: 0.6882, Loss: 0.3275
Epoch   2 Batch  505/1077 - Train Accuracy: 0.7675, Validation Accuracy: 0.6847, Loss: 0.2917
Epoch   2 Batch  510/1077 - Train Accuracy: 0.7066, Validation Accuracy: 0.6747, Loss: 0.3143
Epoch   2 Batch  515/1077 - Train Accuracy: 0.7105, Validation Accuracy: 0.6942, Loss: 0.3237
Epoch   2 Batch  520/1077 - Train Accuracy: 0.7660, Validation Accuracy: 0.6893, Loss: 0.2946
Epoch   2 Batch  525/1077 - Train Accuracy: 0.7078, Validation Accuracy: 0.6903, Loss: 0.3116
Epoch   2 Batch  530/1077 - Train Accuracy: 0.6625, Validation Accuracy: 0.6708, Loss: 0.3296
Epoch   2 Batch  535/1077 - Train Accuracy: 0.7289, Validation Accuracy: 0.6658, Loss: 0.3150
Epoch   2 Batch  540/1077 - Train Accuracy: 0.7258, Validation Accuracy: 0.6875, Loss: 0.2948
Epoch   2 Batch  545/1077 - Train Accuracy: 0.7336, Validati

Epoch   2 Batch  935/1077 - Train Accuracy: 0.7504, Validation Accuracy: 0.7045, Loss: 0.2926
Epoch   2 Batch  940/1077 - Train Accuracy: 0.7562, Validation Accuracy: 0.7482, Loss: 0.2773
Epoch   2 Batch  945/1077 - Train Accuracy: 0.7879, Validation Accuracy: 0.7298, Loss: 0.2793
Epoch   2 Batch  950/1077 - Train Accuracy: 0.7619, Validation Accuracy: 0.7351, Loss: 0.2635
Epoch   2 Batch  955/1077 - Train Accuracy: 0.7531, Validation Accuracy: 0.7464, Loss: 0.2752
Epoch   2 Batch  960/1077 - Train Accuracy: 0.7530, Validation Accuracy: 0.7141, Loss: 0.2680
Epoch   2 Batch  965/1077 - Train Accuracy: 0.7270, Validation Accuracy: 0.7401, Loss: 0.2993
Epoch   2 Batch  970/1077 - Train Accuracy: 0.7480, Validation Accuracy: 0.6935, Loss: 0.2990
Epoch   2 Batch  975/1077 - Train Accuracy: 0.7597, Validation Accuracy: 0.7333, Loss: 0.2746
Epoch   2 Batch  980/1077 - Train Accuracy: 0.7762, Validation Accuracy: 0.7514, Loss: 0.2822
Epoch   2 Batch  985/1077 - Train Accuracy: 0.7934, Validati

Epoch   3 Batch  300/1077 - Train Accuracy: 0.8195, Validation Accuracy: 0.8129, Loss: 0.2378
Epoch   3 Batch  305/1077 - Train Accuracy: 0.8141, Validation Accuracy: 0.7635, Loss: 0.2381
Epoch   3 Batch  310/1077 - Train Accuracy: 0.7598, Validation Accuracy: 0.7859, Loss: 0.2666
Epoch   3 Batch  315/1077 - Train Accuracy: 0.8307, Validation Accuracy: 0.7915, Loss: 0.2297
Epoch   3 Batch  320/1077 - Train Accuracy: 0.7984, Validation Accuracy: 0.7773, Loss: 0.2573
Epoch   3 Batch  325/1077 - Train Accuracy: 0.8110, Validation Accuracy: 0.7607, Loss: 0.2378
Epoch   3 Batch  330/1077 - Train Accuracy: 0.8086, Validation Accuracy: 0.7713, Loss: 0.2468
Epoch   3 Batch  335/1077 - Train Accuracy: 0.8051, Validation Accuracy: 0.7621, Loss: 0.2316
Epoch   3 Batch  340/1077 - Train Accuracy: 0.8289, Validation Accuracy: 0.7798, Loss: 0.2523
Epoch   3 Batch  345/1077 - Train Accuracy: 0.7954, Validation Accuracy: 0.7685, Loss: 0.2222
Epoch   3 Batch  350/1077 - Train Accuracy: 0.8102, Validati

Epoch   3 Batch  740/1077 - Train Accuracy: 0.8496, Validation Accuracy: 0.8200, Loss: 0.2009
Epoch   3 Batch  745/1077 - Train Accuracy: 0.8121, Validation Accuracy: 0.8260, Loss: 0.2168
Epoch   3 Batch  750/1077 - Train Accuracy: 0.8309, Validation Accuracy: 0.8061, Loss: 0.2076
Epoch   3 Batch  755/1077 - Train Accuracy: 0.8313, Validation Accuracy: 0.8082, Loss: 0.2166
Epoch   3 Batch  760/1077 - Train Accuracy: 0.8445, Validation Accuracy: 0.8210, Loss: 0.2118
Epoch   3 Batch  765/1077 - Train Accuracy: 0.8613, Validation Accuracy: 0.8104, Loss: 0.2065
Epoch   3 Batch  770/1077 - Train Accuracy: 0.8233, Validation Accuracy: 0.8366, Loss: 0.2048
Epoch   3 Batch  775/1077 - Train Accuracy: 0.7863, Validation Accuracy: 0.7972, Loss: 0.2083
Epoch   3 Batch  780/1077 - Train Accuracy: 0.8102, Validation Accuracy: 0.7990, Loss: 0.2365
Epoch   3 Batch  785/1077 - Train Accuracy: 0.8374, Validation Accuracy: 0.8008, Loss: 0.1990
Epoch   3 Batch  790/1077 - Train Accuracy: 0.7648, Validati

Epoch   4 Batch  105/1077 - Train Accuracy: 0.8559, Validation Accuracy: 0.8267, Loss: 0.1760
Epoch   4 Batch  110/1077 - Train Accuracy: 0.8680, Validation Accuracy: 0.8217, Loss: 0.1719
Epoch   4 Batch  115/1077 - Train Accuracy: 0.7902, Validation Accuracy: 0.8402, Loss: 0.1919
Epoch   4 Batch  120/1077 - Train Accuracy: 0.8254, Validation Accuracy: 0.8374, Loss: 0.2082
Epoch   4 Batch  125/1077 - Train Accuracy: 0.8690, Validation Accuracy: 0.8366, Loss: 0.1860
Epoch   4 Batch  130/1077 - Train Accuracy: 0.8568, Validation Accuracy: 0.8168, Loss: 0.1739
Epoch   4 Batch  135/1077 - Train Accuracy: 0.8561, Validation Accuracy: 0.8448, Loss: 0.1816
Epoch   4 Batch  140/1077 - Train Accuracy: 0.8836, Validation Accuracy: 0.8249, Loss: 0.1957
Epoch   4 Batch  145/1077 - Train Accuracy: 0.8562, Validation Accuracy: 0.8398, Loss: 0.1895
Epoch   4 Batch  150/1077 - Train Accuracy: 0.8467, Validation Accuracy: 0.8189, Loss: 0.1905
Epoch   4 Batch  155/1077 - Train Accuracy: 0.8719, Validati

Epoch   4 Batch  545/1077 - Train Accuracy: 0.8410, Validation Accuracy: 0.8324, Loss: 0.1819
Epoch   4 Batch  550/1077 - Train Accuracy: 0.8074, Validation Accuracy: 0.8402, Loss: 0.1829
Epoch   4 Batch  555/1077 - Train Accuracy: 0.8742, Validation Accuracy: 0.8441, Loss: 0.1722
Epoch   4 Batch  560/1077 - Train Accuracy: 0.8313, Validation Accuracy: 0.8274, Loss: 0.1652
Epoch   4 Batch  565/1077 - Train Accuracy: 0.8512, Validation Accuracy: 0.8299, Loss: 0.1726
Epoch   4 Batch  570/1077 - Train Accuracy: 0.8322, Validation Accuracy: 0.8345, Loss: 0.1800
Epoch   4 Batch  575/1077 - Train Accuracy: 0.8430, Validation Accuracy: 0.8434, Loss: 0.1570
Epoch   4 Batch  580/1077 - Train Accuracy: 0.8627, Validation Accuracy: 0.8462, Loss: 0.1391
Epoch   4 Batch  585/1077 - Train Accuracy: 0.8757, Validation Accuracy: 0.8580, Loss: 0.1478
Epoch   4 Batch  590/1077 - Train Accuracy: 0.8220, Validation Accuracy: 0.8590, Loss: 0.1821
Epoch   4 Batch  595/1077 - Train Accuracy: 0.8844, Validati

Epoch   4 Batch  985/1077 - Train Accuracy: 0.9047, Validation Accuracy: 0.8540, Loss: 0.1386
Epoch   4 Batch  990/1077 - Train Accuracy: 0.8532, Validation Accuracy: 0.8494, Loss: 0.1668
Epoch   4 Batch  995/1077 - Train Accuracy: 0.8903, Validation Accuracy: 0.8651, Loss: 0.1373
Epoch   4 Batch 1000/1077 - Train Accuracy: 0.8471, Validation Accuracy: 0.8484, Loss: 0.1412
Epoch   4 Batch 1005/1077 - Train Accuracy: 0.8758, Validation Accuracy: 0.8729, Loss: 0.1428
Epoch   4 Batch 1010/1077 - Train Accuracy: 0.8809, Validation Accuracy: 0.8619, Loss: 0.1497
Epoch   4 Batch 1015/1077 - Train Accuracy: 0.8484, Validation Accuracy: 0.8793, Loss: 0.1718
Epoch   4 Batch 1020/1077 - Train Accuracy: 0.9039, Validation Accuracy: 0.8406, Loss: 0.1403
Epoch   4 Batch 1025/1077 - Train Accuracy: 0.8333, Validation Accuracy: 0.8871, Loss: 0.1509
Epoch   4 Batch 1030/1077 - Train Accuracy: 0.8543, Validation Accuracy: 0.8594, Loss: 0.1598
Epoch   4 Batch 1035/1077 - Train Accuracy: 0.9170, Validati

Epoch   5 Batch  350/1077 - Train Accuracy: 0.8758, Validation Accuracy: 0.8512, Loss: 0.1389
Epoch   5 Batch  355/1077 - Train Accuracy: 0.8865, Validation Accuracy: 0.8647, Loss: 0.1287
Epoch   5 Batch  360/1077 - Train Accuracy: 0.8852, Validation Accuracy: 0.8480, Loss: 0.1318
Epoch   5 Batch  365/1077 - Train Accuracy: 0.8738, Validation Accuracy: 0.8512, Loss: 0.1229
Epoch   5 Batch  370/1077 - Train Accuracy: 0.9111, Validation Accuracy: 0.8675, Loss: 0.1251
Epoch   5 Batch  375/1077 - Train Accuracy: 0.8888, Validation Accuracy: 0.8462, Loss: 0.1213
Epoch   5 Batch  380/1077 - Train Accuracy: 0.8887, Validation Accuracy: 0.8540, Loss: 0.1307
Epoch   5 Batch  385/1077 - Train Accuracy: 0.8812, Validation Accuracy: 0.8580, Loss: 0.1209
Epoch   5 Batch  390/1077 - Train Accuracy: 0.8215, Validation Accuracy: 0.8615, Loss: 0.1612
Epoch   5 Batch  395/1077 - Train Accuracy: 0.8750, Validation Accuracy: 0.8686, Loss: 0.1237
Epoch   5 Batch  400/1077 - Train Accuracy: 0.8723, Validati

Epoch   5 Batch  790/1077 - Train Accuracy: 0.8047, Validation Accuracy: 0.8881, Loss: 0.1343
Epoch   5 Batch  795/1077 - Train Accuracy: 0.8867, Validation Accuracy: 0.8839, Loss: 0.1329
Epoch   5 Batch  800/1077 - Train Accuracy: 0.9023, Validation Accuracy: 0.8917, Loss: 0.1251
Epoch   5 Batch  805/1077 - Train Accuracy: 0.8527, Validation Accuracy: 0.8920, Loss: 0.1224
Epoch   5 Batch  810/1077 - Train Accuracy: 0.8538, Validation Accuracy: 0.9009, Loss: 0.1129
Epoch   5 Batch  815/1077 - Train Accuracy: 0.8691, Validation Accuracy: 0.8967, Loss: 0.1290
Epoch   5 Batch  820/1077 - Train Accuracy: 0.8422, Validation Accuracy: 0.8885, Loss: 0.1446
Epoch   5 Batch  825/1077 - Train Accuracy: 0.9031, Validation Accuracy: 0.8683, Loss: 0.1191
Epoch   5 Batch  830/1077 - Train Accuracy: 0.8340, Validation Accuracy: 0.8995, Loss: 0.1374
Epoch   5 Batch  835/1077 - Train Accuracy: 0.8883, Validation Accuracy: 0.8754, Loss: 0.1331
Epoch   5 Batch  840/1077 - Train Accuracy: 0.8621, Validati

Epoch   6 Batch  155/1077 - Train Accuracy: 0.9211, Validation Accuracy: 0.8846, Loss: 0.1137
Epoch   6 Batch  160/1077 - Train Accuracy: 0.8965, Validation Accuracy: 0.8629, Loss: 0.1095
Epoch   6 Batch  165/1077 - Train Accuracy: 0.8598, Validation Accuracy: 0.8608, Loss: 0.1250
Epoch   6 Batch  170/1077 - Train Accuracy: 0.8598, Validation Accuracy: 0.8537, Loss: 0.1171
Epoch   6 Batch  175/1077 - Train Accuracy: 0.8918, Validation Accuracy: 0.8817, Loss: 0.1165
Epoch   6 Batch  180/1077 - Train Accuracy: 0.8969, Validation Accuracy: 0.8778, Loss: 0.1025
Epoch   6 Batch  185/1077 - Train Accuracy: 0.8770, Validation Accuracy: 0.8679, Loss: 0.1137
Epoch   6 Batch  190/1077 - Train Accuracy: 0.9004, Validation Accuracy: 0.8881, Loss: 0.1128
Epoch   6 Batch  195/1077 - Train Accuracy: 0.8801, Validation Accuracy: 0.8857, Loss: 0.1170
Epoch   6 Batch  200/1077 - Train Accuracy: 0.8645, Validation Accuracy: 0.8729, Loss: 0.1270
Epoch   6 Batch  205/1077 - Train Accuracy: 0.8547, Validati

Epoch   6 Batch  595/1077 - Train Accuracy: 0.9273, Validation Accuracy: 0.8899, Loss: 0.0991
Epoch   6 Batch  600/1077 - Train Accuracy: 0.8929, Validation Accuracy: 0.8899, Loss: 0.1041
Epoch   6 Batch  605/1077 - Train Accuracy: 0.9079, Validation Accuracy: 0.8874, Loss: 0.1265
Epoch   6 Batch  610/1077 - Train Accuracy: 0.8993, Validation Accuracy: 0.8945, Loss: 0.1056
Epoch   6 Batch  615/1077 - Train Accuracy: 0.8793, Validation Accuracy: 0.8924, Loss: 0.1009
Epoch   6 Batch  620/1077 - Train Accuracy: 0.9070, Validation Accuracy: 0.8626, Loss: 0.1079
Epoch   6 Batch  625/1077 - Train Accuracy: 0.9109, Validation Accuracy: 0.8629, Loss: 0.1050
Epoch   6 Batch  630/1077 - Train Accuracy: 0.9160, Validation Accuracy: 0.8786, Loss: 0.1153
Epoch   6 Batch  635/1077 - Train Accuracy: 0.8935, Validation Accuracy: 0.9006, Loss: 0.1120
Epoch   6 Batch  640/1077 - Train Accuracy: 0.8854, Validation Accuracy: 0.8704, Loss: 0.0969
Epoch   6 Batch  645/1077 - Train Accuracy: 0.8966, Validati

Epoch   6 Batch 1035/1077 - Train Accuracy: 0.9301, Validation Accuracy: 0.8970, Loss: 0.0887
Epoch   6 Batch 1040/1077 - Train Accuracy: 0.9042, Validation Accuracy: 0.8903, Loss: 0.1012
Epoch   6 Batch 1045/1077 - Train Accuracy: 0.8953, Validation Accuracy: 0.8903, Loss: 0.1169
Epoch   6 Batch 1050/1077 - Train Accuracy: 0.9070, Validation Accuracy: 0.8853, Loss: 0.0943
Epoch   6 Batch 1055/1077 - Train Accuracy: 0.8867, Validation Accuracy: 0.8903, Loss: 0.1134
Epoch   6 Batch 1060/1077 - Train Accuracy: 0.9211, Validation Accuracy: 0.8842, Loss: 0.0772
Epoch   6 Batch 1065/1077 - Train Accuracy: 0.9133, Validation Accuracy: 0.9027, Loss: 0.0943
Epoch   6 Batch 1070/1077 - Train Accuracy: 0.8746, Validation Accuracy: 0.8754, Loss: 0.0911
Epoch   6 Batch 1075/1077 - Train Accuracy: 0.8960, Validation Accuracy: 0.8725, Loss: 0.1053
Epoch   7 Batch    5/1077 - Train Accuracy: 0.8965, Validation Accuracy: 0.8828, Loss: 0.1275
Epoch   7 Batch   10/1077 - Train Accuracy: 0.9071, Validati

Epoch   7 Batch  400/1077 - Train Accuracy: 0.8809, Validation Accuracy: 0.8690, Loss: 0.0945
Epoch   7 Batch  405/1077 - Train Accuracy: 0.8799, Validation Accuracy: 0.8864, Loss: 0.0949
Epoch   7 Batch  410/1077 - Train Accuracy: 0.8676, Validation Accuracy: 0.8842, Loss: 0.1117
Epoch   7 Batch  415/1077 - Train Accuracy: 0.8906, Validation Accuracy: 0.8885, Loss: 0.1052
Epoch   7 Batch  420/1077 - Train Accuracy: 0.9301, Validation Accuracy: 0.8917, Loss: 0.0737
Epoch   7 Batch  425/1077 - Train Accuracy: 0.8813, Validation Accuracy: 0.8789, Loss: 0.0771
Epoch   7 Batch  430/1077 - Train Accuracy: 0.8941, Validation Accuracy: 0.8864, Loss: 0.0998
Epoch   7 Batch  435/1077 - Train Accuracy: 0.9260, Validation Accuracy: 0.8832, Loss: 0.1185
Epoch   7 Batch  440/1077 - Train Accuracy: 0.8555, Validation Accuracy: 0.8725, Loss: 0.1169
Epoch   7 Batch  445/1077 - Train Accuracy: 0.8577, Validation Accuracy: 0.8725, Loss: 0.0965
Epoch   7 Batch  450/1077 - Train Accuracy: 0.9020, Validati

Epoch   7 Batch  840/1077 - Train Accuracy: 0.8953, Validation Accuracy: 0.9055, Loss: 0.0922
Epoch   7 Batch  845/1077 - Train Accuracy: 0.8945, Validation Accuracy: 0.8981, Loss: 0.0782
Epoch   7 Batch  850/1077 - Train Accuracy: 0.8750, Validation Accuracy: 0.8991, Loss: 0.1128
Epoch   7 Batch  855/1077 - Train Accuracy: 0.9152, Validation Accuracy: 0.9080, Loss: 0.1027
Epoch   7 Batch  860/1077 - Train Accuracy: 0.9085, Validation Accuracy: 0.9123, Loss: 0.0900
Epoch   7 Batch  865/1077 - Train Accuracy: 0.9212, Validation Accuracy: 0.8960, Loss: 0.0872
Epoch   7 Batch  870/1077 - Train Accuracy: 0.8540, Validation Accuracy: 0.8878, Loss: 0.1045
Epoch   7 Batch  875/1077 - Train Accuracy: 0.9012, Validation Accuracy: 0.9091, Loss: 0.0983
Epoch   7 Batch  880/1077 - Train Accuracy: 0.8961, Validation Accuracy: 0.9013, Loss: 0.1006
Epoch   7 Batch  885/1077 - Train Accuracy: 0.9112, Validation Accuracy: 0.8913, Loss: 0.0678
Epoch   7 Batch  890/1077 - Train Accuracy: 0.8988, Validati

Epoch   8 Batch  205/1077 - Train Accuracy: 0.8926, Validation Accuracy: 0.8842, Loss: 0.0817
Epoch   8 Batch  210/1077 - Train Accuracy: 0.8981, Validation Accuracy: 0.8935, Loss: 0.0904
Epoch   8 Batch  215/1077 - Train Accuracy: 0.8633, Validation Accuracy: 0.9045, Loss: 0.1032
Epoch   8 Batch  220/1077 - Train Accuracy: 0.9371, Validation Accuracy: 0.9084, Loss: 0.0776
Epoch   8 Batch  225/1077 - Train Accuracy: 0.9098, Validation Accuracy: 0.9109, Loss: 0.0922
Epoch   8 Batch  230/1077 - Train Accuracy: 0.9100, Validation Accuracy: 0.9006, Loss: 0.0721
Epoch   8 Batch  235/1077 - Train Accuracy: 0.8597, Validation Accuracy: 0.9094, Loss: 0.0776
Epoch   8 Batch  240/1077 - Train Accuracy: 0.9246, Validation Accuracy: 0.9070, Loss: 0.0831
Epoch   8 Batch  245/1077 - Train Accuracy: 0.9044, Validation Accuracy: 0.8931, Loss: 0.0782
Epoch   8 Batch  250/1077 - Train Accuracy: 0.9265, Validation Accuracy: 0.8857, Loss: 0.0792
Epoch   8 Batch  255/1077 - Train Accuracy: 0.9062, Validati

Epoch   8 Batch  645/1077 - Train Accuracy: 0.9018, Validation Accuracy: 0.8960, Loss: 0.0843
Epoch   8 Batch  650/1077 - Train Accuracy: 0.9078, Validation Accuracy: 0.9041, Loss: 0.0780
Epoch   8 Batch  655/1077 - Train Accuracy: 0.9074, Validation Accuracy: 0.9098, Loss: 0.0879
Epoch   8 Batch  660/1077 - Train Accuracy: 0.9172, Validation Accuracy: 0.9130, Loss: 0.0802
Epoch   8 Batch  665/1077 - Train Accuracy: 0.9078, Validation Accuracy: 0.9173, Loss: 0.0885
Epoch   8 Batch  670/1077 - Train Accuracy: 0.9183, Validation Accuracy: 0.9183, Loss: 0.0711
Epoch   8 Batch  675/1077 - Train Accuracy: 0.9070, Validation Accuracy: 0.9180, Loss: 0.0817
Epoch   8 Batch  680/1077 - Train Accuracy: 0.8813, Validation Accuracy: 0.9272, Loss: 0.0685
Epoch   8 Batch  685/1077 - Train Accuracy: 0.8945, Validation Accuracy: 0.9173, Loss: 0.0766
Epoch   8 Batch  690/1077 - Train Accuracy: 0.9086, Validation Accuracy: 0.9261, Loss: 0.0825
Epoch   8 Batch  695/1077 - Train Accuracy: 0.9441, Validati

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [79]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [80]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [87]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    unkonwn_int = vocab_to_int["<UNK>"]
    
    int_word = lambda word : vocab_to_int[word] if word in vocab_to_int else unkonwn_int
    
    return [int_word(word) for word in sentence.split()]

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [92]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [150, 56, 134, 63, 181, 108, 221]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [202, 135, 320, 25, 272, 125, 207, 86, 1]
  French Words: il a vu un camion noir rouillé . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.